packages/redis/src/RedisStreamSource.ts

```diff
@@ -0,0 +1,147 @@
+import { RedisOpenTracingTagKeys } from "./RedisClient";
+
+export class RedisStreamSource implements IInputSource, IRequireInitialization, IDisposable {
+    private done: boolean = false;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private spanOperationName: string = "Redis Input Source Client Call";
+    private consumerId: string;
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+        this.consumerId = generate();
+    }
+
+    public async *start(): AsyncIterableIterator<MessageRef> {
+        while (!this.done) {
+            const span = this.tracer.startSpan(this.spanOperationName);
+
+            this.spanLogAndSetTags(
+                span,
+                this.config.db,
+                this.config.readStream,
+                this.config.consumerGroup
+            );
+
+            const pendingMessages = await this.client.xPending(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                5
+            );
+
+            const expiredIdleMessages = pendingMessages.filter(
+                ([, , idleTime]) => idleTime > this.config.idleTimeoutMs
+            );
+
+            if (expiredIdleMessages.length > 0) {
+                // drain expired idle messages
+                for (let [streamId] of expiredIdleMessages) {
+                    const [, msg] = await this.client.xClaim(
+                        span.context(),
+                        MessageRef.name,
+                        this.config.readStream,
+                        this.config.consumerGroup,
+                        this.consumerId,
+                        this.config.idleTimeoutMs,
+                        streamId
+                    );
+
+                    msg.once("released", async () => {
+                        await this.client.xAck(
+                            span.context(),
+                            this.config.readStream,
+                            this.config.consumerGroup,
+                            streamId
+                        );
+                    });
+
+                    yield msg;
+                }
+            } else {
+                const [streamId, msg] = await this.client.xReadGroupObject<MessageRef>(
+                    span.context(),
+                    MessageRef.name,
+                    this.config.readStream,
+                    this.config.consumerGroup,
+                    this.consumerId
+                );
+
+                msg.once("released", async () => {
+                    await this.client.xAck(
+                        span.context(),
+                        this.config.readStream,
+                        this.config.consumerGroup,
+                        streamId
+                    );
+                });
+
+                yield msg;
+            }
+
+            span.finish();
```

Comment:
think you need to finish the span in the `msg.once("released")` callback so it properly reflects the time spent processing the message end-to-end